dependencydiff/dependencydiff.go

```diff
@@ -0,0 +1,119 @@
+// Copyright 2022 Security Scorecard Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dependencydiff
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"path"
+
+	"github.com/ossf/scorecard/v4/checks"
+	"github.com/ossf/scorecard/v4/clients"
+	"github.com/ossf/scorecard/v4/clients/githubrepo"
+	"github.com/ossf/scorecard/v4/clients/githubrepo/roundtripper"
+	"github.com/ossf/scorecard/v4/pkg"
+)
+
+// GetDependencyDiffResults gets dependency changes between two given code commits BASE and HEAD
+// along with the Scorecard check results of the dependencies, and returns a slice of DependencyCheckResult.
+func GetDependencyDiffResults(ownerName, repoName, baseSHA, headSHA string) ([]pkg.DependencyCheckResult, error) {
+	ctx := context.Background()
+	// Fetch dependency diffs using the GitHub Dependency Review API.
+	deps, err := FetchDependencyDiffData(ctx, ownerName, repoName, baseSHA, headSHA)
+	if err != nil {
+		return nil, fmt.Errorf("error fetching dependency changes: %w", err)
+	}
+
+	ghRepo, err := githubrepo.MakeGithubRepo(path.Join(ownerName, repoName))
+	if err != nil {
+		return nil, fmt.Errorf("error creating the github repo: %w", err)
+	}
+	ghRepoClient := githubrepo.CreateGithubRepoClient(ctx, nil)
+	ossFuzzRepoClient, err := githubrepo.CreateOssFuzzRepoClient(ctx, nil)
+	if err != nil {
+		return nil, fmt.Errorf("error creating the oss fuzz repo client: %w", err)
+	}
+	vulnsClient := clients.DefaultVulnerabilitiesClient()
+
+	results := []pkg.DependencyCheckResult{}
+	for _, d := range deps {
+		if d.SourceRepository != nil {
+			if *d.SourceRepository == "" {
+				continue
+			}
+			// Running scorecard on this dependency and fetch the result.
+			// TODO: use the Scorecare REST API to retrieve the Scorecard result statelessly.
+			scorecardResult, err := pkg.RunScorecards(
+				ctx,
+				ghRepo,
+				clients.HeadSHA,
+				checks.AllChecks,
+				ghRepoClient,
+				ossFuzzRepoClient,
+				nil,
+				vulnsClient,
+			)
+			if err != nil {
+				return nil, fmt.Errorf("error fetching the scorecard result: %w", err)
+			}
+			depCheckResult := pkg.DependencyCheckResult{
```

Comment:
see comment above: we may want to append a result even if the scorecard result is `nil`. 
I'm not sure what we should do if scorecard run returns an error: I'm inclined to not fail immediately and instead return the scorecard return error. This way even if one run fails, it still provides the results for other dependencies.

Reply:
I agree that we can return the scorecard error and continue to the next dependency if the run fails. 

When it fails, what will the function `RunScorecards`-returned `ScorecardResult` be like? Can I append the failed `scorecardResult` directly to the results, or should I generate another error message using another module?

Reply:
Scorecard results should have an `error` field we can use to populate it. Can you check?

Reply:
sorry for not explaining my idea very clearly.

i think that's possible. i can use the error returned by `RunScorecards` to decide whether:
(1) it succeeds with `err==nil` (the situation when my API returns the result normally), or 
(2) it fails with `err!=nil` (my API returns an empty result for this dependency rather than directly returns and skip all the following dependencies).